ddl/column.go

```diff
@@ -1054,37 +1054,34 @@ func (w *worker) doModifyColumnTypeWithData(
 		// Clean up the channel of notifyCancelReorgJob. Make sure it can't affect other jobs.
 		w.reorgCtx.cleanNotifyReorgCancel()
 
-		// Remove the old column and indexes. Update the relative column name and index names.
-		oldIdxIDs := make([]int64, 0, len(changingIdxs))
-		tblInfo.Columns = tblInfo.Columns[:len(tblInfo.Columns)-1]
-		for _, cIdx := range changingIdxs {
-			idxName := getChangingIndexOriginName(cIdx)
-			for i, idx := range tblInfo.Indices {
-				if strings.EqualFold(idxName, idx.Name.O) {
-					cIdx.Name = model.NewCIStr(idxName)
-					tblInfo.Indices[i] = cIdx
-					oldIdxIDs = append(oldIdxIDs, idx.ID)
-					break
-				}
-			}
-		}
-		changingColumnUniqueName := changingCol.Name
-		changingCol.Name = colName
-		changingCol.ChangeStateInfo = nil
-		// After changing the column, the column's type is change, so it needs to set OriginDefaultValue back
-		// so that there is no error in getting the default value from OriginDefaultValue.
-		// Besides, nil data that was not backfilled in the "add column" is backfilled after the column is changed.
-		// So it can set OriginDefaultValue to nil.
-		if err = changingCol.SetOriginDefaultValue(nil); err != nil {
+		oldIdxIDs := getOldIndexesID(tblInfo, oldCol) // used by GC delete range.
+
+		internalColName := changingCol.Name
+		changingCol, err = replaceOldColumn(tblInfo, oldCol.ID, changingCol.ID, colName)
+		if err != nil {
+			job.State = model.JobStateRollingback
 			return ver, errors.Trace(err)
 		}
-		tblInfo.Indices = tblInfo.Indices[:len(tblInfo.Indices)-len(changingIdxs)]
-		// Adjust table column offset.
-		if err = adjustColumnInfoInModifyColumn(job, tblInfo, changingCol, oldCol, pos, changingColumnUniqueName.L); err != nil {
```

Comment:
Could we adjust the function of `adjustColumnInfoInModifyColumn`？Then we can still call it here.